### PR TITLE
Implement Stochastic Rounding

### PR DESCRIPTION
**Pull request by crcrpar**
This PR introduces Stochastic Rounding (SR). We've had success with SR training networks to convergence using only FP16 weights. We've recently confirmed with this implementation that ResNet50 and Transformer can be trained in FP16 without significant loss. We have also previously trained a wider set of networks maintaining convergence behavior. Runs were compared with deprecated nvidia/apex's AMP using FP32 "master weights".
Using SR prevents the need to have an FP32 copy of the weights. This can significantly decrease the memory footprint of the network.

## Recent results:
* ResNet50 and SGD-SR converged to 76 of top-1
Trained with native AMP features (GradScaler and autocast)
Parameters are in FP16 by calling model.half()
* FP16 Transformer and Adam-SR at scale reached BLEU score of 25.0 with the same number of epochs as with non-SR Adam.

## How Stochastic Rounding Works
It rounds FP32 values to FP16 stochastically though we can implement rounding algorithms to dtypes other than FP16. Assume `x` is an FP32, `y` and `z` are FP16 and 0 < `y` < `x` < `z`. Usually, `x` is rounded to `y` if (1) round_to_zero or (2) round_to_nearest and abs(x - y) < abs(x - z). However, SR rounds `x` to `y` (`z`) with the probability of abs(x - z) / abs(y - z) (abs(x - y) / abs(y - z)).

## Design of `round_stochastically` kernel
Current implementation only supports the cast from FP32/64 to FP16, however, we can support cast to other dtypes like bfloat16.

## Design of Optimizer kernels
The kernels load params, gradients, and some momentums in `scalar_t` and cast them into FP32 (`float`) before do some update math in 32 bits. After that, they write updates values in `scalar_t` and SR is applied in this writing out phase. These custom optimizers, even if they will not be able to be in `torch.optim`, works better than `torch.optim` optimizers in low precision training.

### CODE REVIEW

**Comment by crcrpar** [comment]
@vincentqb Could you review this as you are the governor of `torch.optim`?

**Comment by ngimel** [comment]
This PR adds operations with questionable behavior (hard-coded datatypes, supported only on GPU) to public APIs, which is bad. 
cc @vincentqb for the new optimizers introduced, I'm not sure that we need these new optimizer types, and in any case I don't think we can have optimizers that work only on CUDA tensors.

**Comment by ngimel** [comment]
A similar narrowly-taylored optimizer PR was not merged last year https://github.com/pytorch/pytorch/pull/23522, and I believe most of the concerns expressed there are valid for this PR as well. 
This PR 
1) is opened out of the blue, without filing an issue to propose adding new features/optimizers
2) does not describe in sufficient detail how stochastic rounding works and why is it better than keeping copies of master parameters for low-precision training
3) provides stochastic rounding variants for sgd with momentum and Adam, by creating new optimizers. It essentially freezes the implementation, makes it unhackable, and works only on CUDA. Is the plan to provide similar option to other optimizers? Other backends, CPU at least? If not, why not?
4) As I've already mentioned, adds publicly-exposed APIs that work only for CUDA tensors. 
This seems like a good feature to have in add-on packages for pytorch, but having it as a core feature would require a much better planning. 
cc'ing active participants on #23522
@soumith @dzhulgakov @SsnL

**Comment by soumith** [comment]
nothing that @ngimel said sound incorrect to me.

The PR might be a conversation starter on what parts make sense to include, or what library calls to add, along with the return on investment, and then develop a checklist (like CPU support) and plan. The PR is not ready to be merged in it's current form for sure.

**Comment by csarofeen** [comment]
@soumith understandable, but these are exactly the things we were seeking guidance on by submitting this PR.

Stochastic rounding is an approach that can reduce the memory footprint of network parameters from 1.5x fp32 training to 0.5x for mixed precision training. This has been a consistent source of frustration for many users, as the extra footprint can be unexpected (fp16 should take half the memory, right?).  If we can decrease the footprint, extra perf can frequently be gained by increasing batch size. This is an approach we've generally been interested in exploring and even something that Natalia and Ruberry pushed in another lifetime.

I agree it's extremely difficult to add this optimizer in a consistent pytorch manner, however, we worked on this PR to explore what those challenges are and wanted to post it to ask for guidance and recommendations.

I believe this is something that makes sense to work on together and understand it could be a long term project. However, please let me know if this is something you think is too difficult and best seen only in Apex.

**Comment by vincentqb** [comment]
Thanks @crcrpar for starting the discussion. I second @ngimel and @soumith. This needs quite a bit of discussion, and this PR cannot be merged in its current form. In the future, I suggest opening an issue before a PR to discuss what is desired and the form it could eventually take.
